tket/src/WeightSubgrMono/Searching/NodeListTraversal.cpp

```diff
@@ -0,0 +1,289 @@
+    VertexWSM t_vertex) {
+  data_for_this_pv.entries[data_for_this_pv.entries_back_index + 1].node_level =
+      raw_data.current_node_level;
+
+  auto& new_domain =
+      data_for_this_pv.entries[data_for_this_pv.entries_back_index + 1].domain;
+  new_domain =
+      data_for_this_pv.entries[data_for_this_pv.entries_back_index].domain;
+  TKET_ASSERT(new_domain.erase(t_vertex) == 1);
+}
+
+// data_for_this_pv.entries_back_index should now be for
+// the new domain we'll create.
+static void complete_move_down_with_resized_vectors_and_indices(
+    NodesRawData& raw_data, NodesRawData::DomainData& data_for_this_pv,
+    VertexWSM p_vertex, VertexWSM t_vertex) {
+  // Remember, the domain data and node data
+  // at the current level is now "junk".
+  auto& new_domain =
+      data_for_this_pv.entries[data_for_this_pv.entries_back_index].domain;
+  new_domain.clear();
+  new_domain.insert(t_vertex);
+
+  data_for_this_pv.entries[data_for_this_pv.entries_back_index].node_level =
+      raw_data.current_node_level;
+
+  auto& new_node_data = raw_data.nodes_data[raw_data.current_node_level];
+  new_node_data.new_assignments.resize(1);
+  new_node_data.new_assignments[0] = {p_vertex, t_vertex};
+  new_node_data.nogood = false;
+
+  new_node_data.scalar_product =
+      raw_data.nodes_data[raw_data.current_node_level - 1].scalar_product;
+
+  new_node_data.total_p_edge_weights =
+      raw_data.nodes_data[raw_data.current_node_level - 1].total_p_edge_weights;
+}
+
+void NodeListTraversal::move_down(VertexWSM p_vertex, VertexWSM t_vertex) {
+  // We can only move down from a valid, fully reduced node.
+  bool existing_node_valid = !m_raw_data.get_current_node().nogood;
+  TKET_ASSERT(existing_node_valid);
+  TKET_ASSERT(m_raw_data.get_current_node().new_assignments.empty());
+
+  auto& data_for_this_pv = m_raw_data.domains_data.at(p_vertex);
+  auto& existing_domain =
+      data_for_this_pv.entries[data_for_this_pv.entries_back_index].domain;
+  auto iter = existing_domain.find(t_vertex);
+
+  // TV must be present!
+  TKET_ASSERT(iter != existing_domain.end());
+  existing_node_valid = when_moving_down_check_current_domain_size(
+      existing_domain, m_raw_data, p_vertex, t_vertex);
+
+  // Next, we need to erase TV from the existing domain.
+  if (existing_node_valid) {
+    if (data_for_this_pv.entries[data_for_this_pv.entries_back_index]
+            .node_level == m_raw_data.current_node_level) {
+      // The data is not shared by previous nodes, so we can just overwrite it
+      // in place.
+      existing_domain.erase(iter);
+
+      // Don't forget the new singleton domain   Dom(PV) = {TV}
+      // we'll create shortly!
+      resize_if_index_is_invalid(
+          data_for_this_pv.entries, data_for_this_pv.entries_back_index + 1);
+
+      ++data_for_this_pv.entries_back_index;
+    } else {
+      resize_if_index_is_invalid(
+          data_for_this_pv.entries, data_for_this_pv.entries_back_index + 2);
```

Comment:
See the comments a bit above " // The data is not shared by previous nodes, so we can just overwrite it
      // in place."

In that case (the "if" clause), we can just change the stored domain, and then add ONE extra domain (because we're moving down).

But we're in the other case ("else" clause): not only do we have a make a COPY of the existing domain and then store that COPY and alter it (copy-on-write: remember that the domain objects are shared between several node levels. Since we're altering only one "logically", we first have to check if it's shared), we ALSO have to make another domain object (since we're moving down, and this is different again; the current domain has to have our PV->TV choice removed, and the domain for the next node level has to be the singleton Dom(PV) = {TV}).

I think I will also wrap this copy-on-write stuff with another template class.